### DISCUSSION TITLE
rp2040 and PCF8574

### DISCUSSION THREAD

**Discussion by dietmarczekay** [earlephilhower/arduino-pico#1227] [Q&A] [answered]
Hi, I didn't get the I2C connection between pico board and pcf8574 from waveshare to work on GPIO0 nd GPIO1.
For intitialisation i use:
```
i2c_init(i2c_default, 100 * 1000);

  gpio_set_function(I2C_SDA, GPIO_FUNC_I2C);
  gpio_set_function(I2C_SCL, GPIO_FUNC_I2C);
  gpio_pull_up(I2C_SDA);
  gpio_pull_up(I2C_SCL);

  Wire.setSDA(0);
  Wire.setSCL(1);
  Wire.begin();
```
sending test data is done by
```
for(int i = 0; i<10;i++){
    Serial.println("Write 0");
    Wire.beginTransmission(address);
    Wire.write(byte(0));
    byte error = Wire.endTransmission();

    if (error == 0) {
      Serial.println("OK");
    } else {
      Serial.println("Fail");
    }
    delay(200);

    Serial.println("Write 1");
    Wire.beginTransmission(address);
    Wire.write(3);
    error = Wire.endTransmission();

    if (error == 0) {
      Serial.println("OK");
    } else {
      Serial.println("Fail");
    }
    delay(200);
  }
```

Any idea what`s wrong?
The SDA and SCL are pulled up by a separate resistor. Using GPIO4 and GPIO5 works correctly (without setSDA and setSCL.

```
  i2c_init(i2c_default, 100 * 1000);
  gpio_set_function(PICO_DEFAULT_I2C_SDA_PIN, GPIO_FUNC_I2C);
  gpio_set_function(PICO_DEFAULT_I2C_SCL_PIN, GPIO_FUNC_I2C);
  gpio_pull_up(PICO_DEFAULT_I2C_SDA_PIN);
  gpio_pull_up(PICO_DEFAULT_I2C_SCL_PIN);
```

Dietmar

**Comment by Andy2No** [comment]
@dietmarczekay Where are I2C_SDA and I2C_SCL defined?   What libraries are included?

**Comment by dietmarczekay** [comment]
both are defined separately

```
#define I2C_SDA 0
#define I2C_SCL 1
#define address 0x21
```
and the following libraries are included:
```
#include <Wire.h>
#include <stdio.h>
#include "pico/stdlib.h"
#include "hardware/i2c.h"
```
the jumper on the pcf8574 board are set to adress 0x21

**Comment by Andy2No** [comment]
Maybe it's this?

https://raspberrypi.github.io/pico-sdk-doxygen/group__hardware__gpio.html#ga2af81373f9f5764ac1a5bd6097477530

"Each GPIO can have one function selected at a time. Likewise, each peripheral input (e.g. UART0 RX) should only be selected on one GPIO at a time. "

That suggests you should free up the I2C peripheral first, by setting the function of the default I2C pins as something else, e.g.  GPIO_FUNC_PIO0 or GPIO_FUNC_PIO1 - I'm not clear which of those two is the normal GPIO pin setting.

https://raspberrypi.github.io/pico-sdk-doxygen/group__hardware__gpio.html

**Comment by dietmarczekay** [comment]
UART ist disabled via platformio settings. And i thought the Pin function ist set by the gpio_set_function command 😐

**Comment by Andy2No** [comment]
> Using GPIO4 and GPIO5 works correctly (without setSDA and setSCL.

You could try:

```
gpio_set_function(4, GPIO_FUNC_NULL);
gpio_set_function(5. GPIO_FUNC_NULL);
```
before:

```
   gpio_set_function(I2C_SDA, GPIO_FUNC_I2C);
   gpio_set_function(I2C_SCL, GPIO_FUNC_I2C);
   gpio_pull_up(I2C_SDA);
   gpio_pull_up(I2C_SCL);
 
   Wire.setSDA(0);
   Wire.setSCL(1);
   Wire.begin();
```

I don't know if this will work, but the idea is to detach pins 4 and 5 from the internal I2C peripheral, before you try to assign the new pins and initialise it.    The document linked to in my previous post suggests that it's possible to have two pins assigned to the same peripheral function, but that you shouldn't do that because it will cause problems.

Setting GPIO_FUNC_NULL will disable those two pins, which might not be what you want, but seems safe to try.

**Comment by dietmarczekay** [comment]
I will give it a try tomorrow. Thanks

**Comment by dietmarczekay** [comment]
No improvement. Do you know any working example?
I first thought, I damaged my PICO with pullup resistors against 5V, But a micropython example works as expected.

**Comment by dietmarczekay** [comment]
```
from machine import I2C, Pin
import utime

LED_PIN = 25
meine = 0
led = machine.Pin(LED_PIN, machine.Pin.OUT)

i2c = I2C(1, scl=Pin(11), sda=Pin(10), freq=100000)
i2c.start
I2C_ADDR = i2c.scan()[0]
print(hex(I2C_ADDR))

while True:
 if meine == 0:
  led.value(1)
  i2c.writeto(I2C_ADDR, b'W')
  print('f')
  meine = 1
 else:
  led.value(0)
  i2c.writeto(I2C_ADDR, b'0')
  print('1')
  meine = 0
 utime.sleep(1.5)
```

**Comment by Andy2No** [comment]
@dietmarczekay I have no experience of platformIO, or any real desire to learn to use it.

Is trying it from just the Arduino IDE an option?  That's all I use with the Pico, with Earle's core added via the Boards manager.

**Comment by dietmarczekay** [comment]
In Arduino the code works as expected:
```

#include <Wire.h>

// I2C address
static const uint8_t I2C_ADDR = 0x21;

static const uint sda_pin = 10;
static const uint scl_pin = 11;

i2c_inst_t *i2c = i2c1;

uint8_t data0 = 0b00000001;
uint8_t data1 = 0b00001111;
int ret2;

static bool reserved_addr(uint8_t addr) {
    return (addr & 0x78) == 0 || (addr & 0x78) == 0x78;
}

void setup() {
  Serial.begin(115200);
  delay(1500);

  gpio_init(PICO_DEFAULT_LED_PIN);
  gpio_set_drive_strength(PICO_DEFAULT_LED_PIN, GPIO_DRIVE_STRENGTH_8MA);
  gpio_set_dir(PICO_DEFAULT_LED_PIN, GPIO_OUT);

  gpio_init(15);
  gpio_set_drive_strength(15, GPIO_DRIVE_STRENGTH_8MA);
  gpio_set_dir(15, GPIO_OUT);
  gpio_put(15, false);
  
  // Initialize I2C pins
  gpio_set_function(sda_pin, GPIO_FUNC_I2C);
  gpio_set_function(scl_pin, GPIO_FUNC_I2C);
  gpio_pull_up(sda_pin);
  gpio_pull_up(scl_pin);


  //Initialize I2C port at 400 kHz
  i2c_init(i2c, 100 * 1000);


  Serial.println("Scanning...");
  for (int addr = 0; addr < (1 << 7); ++addr) {
    if (addr % 16 == 0) {
        Serial.printf("%02x ", addr);
    }

    // Perform a 1-byte dummy read from the probe address. If a slave
    // acknowledges this address, the function returns the number of bytes
    // transferred. If the address byte is ignored, the function returns
    // -1.

    // Skip over any reserved addresses.
    int ret;
    uint8_t rxdata;
    if (reserved_addr(addr))
        ret = PICO_ERROR_GENERIC;
    else
        ret = i2c_read_blocking(i2c, addr, &rxdata, 1, false);
        //ret = i2c_read_blocking(i2c_default, addr, &rxdata, 1, false);

    Serial.print(ret < 0 ? "." : "@");
    Serial.print(addr % 16 == 15 ? "\n" : "  ");
  }
  Serial.print("Test");
  
  ret2 = i2c_write_blocking(i2c, I2C_ADDR, &data0, 1, true);
  
  Serial.println(ret2);

}

void loop() {
  delay(1000);

  gpio_put(PICO_DEFAULT_LED_PIN, true);
  //Serial.print("Test");
}
```

**Comment by dietmarczekay** [comment, marked as answer]
I found a solution for my problem. It seems the problem was, that I didn't used all 8 ports. So 4 of them are floating around. As a result the PCF8574 didn't showed the correct answer to sending a zero (light up all 4 LEDs).
Pulling up the unused ports to 3V3 solved this, so now sending a zero, all LEDs are lighted up.
